### PR TITLE
i18n and language selection for projects

### PR DESCRIPTION
**Pull request by snowyu**
* prompt language internationalization support
  * Note: can not disable it completely yet.
* prompt Which languages do you want to support
* prompt Which language are the default(fall back)
* add the i18n common module for service and language switcher component
  * Note: the language switcher of the bootstrap is not ready.
* Always import common/genernal modules with the mapping path(to rename easily).

Related issue https://github.com/ngx-rocket/generator-ngx-rocket/issues/262

### CODE REVIEW

**Comment by sinedied** [comment]
Thanks for the PR! 
Could you take a look at the lint issues to fix the CI builds?

**Comment by snowyu** [comment]
ok, fixed lint error.

**Comment by bursauxa** [comment]
@snowyu The build is still failing - I noticed that in `_.shell.component.spec.ts`, on the line where you added I18nModule (line36), the comma is not actually a comma, but a similar-looking character (`，` instead of `,`). There may be other issues beyond that, as this issue breaks the build during the compilation stage. Are you still working on this topic?

**Comment by bursauxa** [comment]
There are still several errors on the continuous integration platforms, and several merge conflicts now. 

@snowyu Are you still working on this PR?

**Comment by snowyu** [comment]
@bursauxa almost all error fixed. notice me if you found any problem.

**Comment by bursauxa** [comment]
Well, there are still CI errors (8 failures out of 38 cases on Travis), so that is a problem already.

Generally speaking, I do not look at PRs until they pass CI, because CI errors mean that the code will have to be changed anyways.

**Comment by snowyu** [comment]
Oh, I got it. The new ionic-settings page should use i18n module.

**Comment by snowyu** [comment]
Ok, CI passed now.

File: generators/app/index.js [hunk 1]
```diff
@@ -80,6 +81,13 @@ class NgxGenerator extends Generator {
   }
 
   configuring() {
+    if (!this.props['i18n-langs']) {
+      this.props['i18n-langs'] = ['en-US'];
+    }
+    getLangs().forEach(v => {
+      this._prefixRules[v] = (props => props['i18n-langs'].indexOf(v) >= 0);
```
Comment (bursauxa) [inline generators/app/index.js]: This should be `.includes(...)` rather than the more obscure `.indexOf(...) >= 0`. See a few lines below.

File: generators/app/prompts.js [hunk 2]
```diff
@@ -115,6 +117,26 @@ module.exports = [
...
+    type: 'checkbox',
+    name: 'i18n-langs',
+    message: 'Which languages do you want to support?',
+    when: props => props.i18n,
+    choices: () => getLangs().map(lang => ({value: lang, name: lang, checked: true}))
+  },
+  {
+    type: 'list',
+    name: 'i18n-default',
+    message: 'Which language are the default(fall back)?',
```
Comment (bursauxa) [inline generators/app/prompts.js]: `is` the default rather than `are`. Also, there should be a space between `default` and `(`.

File: generators/app/templates/src/app/i18n/i18n-locales.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export default {
+  'zh-CN': '简体中文',
+  'en-US': 'English(US)',
```
Comment (bursauxa) [inline generators/app/templates/src/app/i18n/i18n-locales.ts]: Add a space before parenthesis.

File: generators/app/templates/src/app/i18n/i18n.service.spec.ts [hunk 4]
```diff
@@ -5,7 +5,8 @@ import { Subject } from 'rxjs';
 import { extract, I18nService } from './i18n.service';
 
 const defaultLanguage = 'en-US';
-const supportedLanguages = ['eo', 'en-US', 'fr-FR'];
+const systemLanguage = 'zh-CN';
```
Comment (bursauxa) [inline generators/app/templates/src/app/i18n/i18n.service.spec.ts]: I do not see where `systemLanguage` is used?

File: generators/app/templates/src/app/i18n/__raw.i18n-lang-selector.component.html [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<select (change)="setLanguage($event.target.value)">
+  <option *ngFor="let language of languages" [value]="language" [selected]="language === currentLanguage">
+    {{language | translate}}
```
Comment (bursauxa) [inline generators/app/templates/src/app/i18n/__raw.i18n-lang-selector.component.html]: From a readability perspective, I believe that entries in a language selection dropdown should not be translated. For instance, if I can not read Chinese, and every entry in the list is written in Chinese, then I will never manage to find my own language in the list.

(this comment applies to all four components)

Comment (snowyu) [reply generators/app/templates/src/app/i18n/__raw.i18n-lang-selector.component.html]: No, the list should display the language name with the localize name.

```html
<div aria-labelledby="language-dropdown" ngbdropdownmenu="" class="dropdown-menu show" x-placement="bottom-right" style="top: 40px; left: -50px;">
    <!--bindings={
  "ng-reflect-ng-for-of": "en-US,fr-FR,zh-CN"
}--><button class="dropdown-item">
      English (US)
    </button><button class="dropdown-item">
      Français
    </button><button class="dropdown-item">
      简体中文
    </button>
  </div>
```

Comment (bursauxa) [reply generators/app/templates/src/app/i18n/__raw.i18n-lang-selector.component.html]: The result is good. I just realized it comes from i18n-locales.ts, so it's all defined in one place with different languages in the same file, which is a godo solution for this use case.

**Comment by bursauxa** [comment]
The Travis failure looks like a false positive to me (at least it looks like it's completely unrelated to this PR). Waiting for a second opinion.

File: generators/app/templates/src/environments/_environment.prod.ts [hunk 6]
```diff
@@ -5,9 +5,6 @@ export const environment = {
   production: true,
   version: env.npm_package_version,
   serverUrl: 'https://api.chucknorris.io',
-  defaultLanguage: 'en-US',
-  supportedLanguages: [
-    'en-US',
-    'fr-FR'
-  ]
+  defaultLanguage: '<%=props['i18n-default'] || (props['i18n-langs'] && props['i18n-langs'][0])%>',
+  supportedLanguages: [<%- (props['i18n-langs'] || []).map( lang => "'"+ lang +"'").join(', ') %>]
```
Comment (sinedied) [inline generators/app/templates/src/environments/_environment.prod.ts]: `'${lang}'` can be used instead of `"'" + lang + "'"` for better readability

File: generators/app/templates/src/environments/_environment.ts [hunk 7]
```diff
@@ -10,9 +10,6 @@ export const environment = {
   production: false,
   version: env.npm_package_version + '-dev',
   serverUrl: '/api',
-  defaultLanguage: 'en-US',
-  supportedLanguages: [
-    'en-US',
-    'fr-FR'
-  ]
+  defaultLanguage: '<%=props['i18n-default'] || (props['i18n-langs'] && props['i18n-langs'][0])%>',
+  supportedLanguages: [<%- (props['i18n-langs'] || []).map( lang => "'"+ lang +"'").join(', ') %>]
```
Comment (sinedied) [inline generators/app/templates/src/environments/_environment.ts]: same as previous comment

File: generators/app/templates/src/app/i18n/__ionic.i18n-lang-selector.component.html [hunk 8]
```diff
@@ -0,0 +1,4 @@
+<button ion-button [attr.round]="round" [attr.outline]="outline" [attr.float-right]="pos === 'right' || null" [attr.float-left] = "pos === 'left' || null"  [color]="color" small [attr.icon-only]="!withText&&withIcon" (click)="changeLanguage()" [title]="'Change language' | translate">
```
Comment (sinedied) [inline generators/app/templates/src/app/i18n/__ionic.i18n-lang-selector.component.html]: missing spaces here: `!withText&&withIcon` => `!withText && withIcon`

File: generators/app/templates/src/translations/__zh-CN._zh-CN.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
+{
+  "APP_NAME": "<%= props.appName %>",
+	"About": "关于",
+	"Hello world !": "你好世界！",
+	"Home": "首页",
+  "Change language": "切换语言",
```
Comment (sinedied) [inline generators/app/templates/src/translations/__zh-CN._zh-CN.json]: indentation issues here

File: generators/app/templates/tsconfig.json [hunk 10]
```diff
@@ -20,6 +20,10 @@
     "baseUrl": "src",
     "paths": {
       "@app/*": ["app/*"],
+      "@i18n": ["app/i18n"],
```
Comment (sinedied) [inline generators/app/templates/tsconfig.json]: As I commented in [your proposal here](https://github.com/ngx-rocket/generator-ngx-rocket/issues/277), I'm ok with adding these as a flag option, but not force it by default.
Anyways, that should be done in a separate PR as it's not directly related to i18n work here, could you change it back for this PR?

File: generators/app/prompts.js [hunk 11]
```diff
@@ -115,6 +117,26 @@ module.exports = [
...
   },
+  {
+    type: 'confirm',
+    name: 'i18n',
+    message: 'Do you want language internationalization support?',
+    default: true
+  },
+  {
+    type: 'checkbox',
+    name: 'i18n-langs',
```
Comment (sinedied) [inline generators/app/prompts.js]: use `i18nLanguages` for consistency (this way you can also use direct dotted notation when accessing props, ie: `props.i18nLanguages`)

File: generators/app/prompts.js [hunk 12]
```diff
@@ -115,6 +117,26 @@ module.exports = [
...
+    message: 'Which languages do you want to support?',
+    when: props => props.i18n,
+    choices: () => getLangs().map(lang => ({value: lang, name: lang, checked: true}))
+  },
+  {
+    type: 'list',
+    name: 'i18n-default',
```
Comment (sinedied) [inline generators/app/prompts.js]: use `i18Default`, for same reason as above

File: generators/app/prompts.js [hunk 13]
```diff
@@ -115,6 +117,26 @@ module.exports = [
     default: false,
     when: props => props.ui !== 'ionic'
   },
+  {
```
Comment (sinedied) [inline generators/app/prompts.js]: Since i18n cannot be completely disabled in the current code base, I think we can remove this prompt for now and only ask which language to support

File: generators/app/prompts.js [hunk 14]
```diff
@@ -115,6 +117,26 @@ module.exports = [
...
+    type: 'confirm',
+    name: 'i18n',
+    message: 'Do you want language internationalization support?',
+    default: true
+  },
+  {
+    type: 'checkbox',
+    name: 'i18n-langs',
+    message: 'Which languages do you want to support?',
```
Comment (sinedied) [inline generators/app/prompts.js]: maybe a validator should be added here to ensure at least 1 language is selected?

**Comment by sinedied** [comment]
I also changed the target of this PR to the upcoming `6.0` release as it's a big change, but also because it will (unfortunately) conflict with https://github.com/ngx-rocket/generator-ngx-rocket/pull/369. Whoever gets its PR to be merged first won't have to deal with it 😝

File: generators/app/templates/_tslint.json [hunk 15]
```diff
@@ -148,7 +148,7 @@
     "component-selector": [
       true,
       "element",
-      "app",
+      ["app", "i18n"],
```
[the remark this answers is not in the document]
Comment (snowyu) [reply generators/app/templates/_tslint.json]: the `app` prefix means only used in app.  `i18n` means  not depend on app.

File: README.md [hunk 16]
```diff
@@ -74,6 +74,7 @@ src/                         project source code
 |- app/                      app components
 |  |- core/                  core module (singleton services and single-use components)
 |  |- shared/                shared module  (common components, directives and pipes)
+|  |- i18n/                  international language service and component
```
[the remark this answers is not in the document]
Comment (snowyu) [reply README.md]: I think the internal third-party modules used by application is normal.

The services and components are easier to reuse if they are organized according to business.

The services which putting into the `core` are the "sigleton" pattern.  But it is not only way to do so.

File: generators/app/prompts.js (hunk 13 above)
Comment (snowyu) [reply generators/app/prompts.js]: I have no idea about this. Currently this is just like a shortcut to choose English language only quickly. 
follow your decision.

File: generators/app/prompts.js (hunk 14 above)
Comment (snowyu) [reply generators/app/prompts.js]: Does it support validator? I've validated it in index.js

```js
  configuring() {
    if (!this.props['i18nLanguages']) {
      this.props['i18nLanguages'] = ['en-US'];
    }
```

File: generators/app/templates/tsconfig.json (hunk 10 above)
Comment (snowyu) [reply generators/app/templates/tsconfig.json]: I put these "third-party" modules (which could be reused) across the application here.

File: generators/app/templates/_tslint.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (bursauxa) [reply generators/app/templates/_tslint.json]: In this case I believe that if you remove i18n, it works exactly the same way (because i18n is a sub-folder of app). So that would be unnecessary code.

File: generators/app/prompts.js (hunk 14 above)
Comment (sinedied) [reply generators/app/prompts.js]: Yes, using the `validate` property, see https://github.com/SBoudrias/Inquirer.js/#objects

I missed you fallback though, so either one is good.

File: generators/app/prompts.js (hunk 13 above)
Comment (sinedied) [reply generators/app/prompts.js]: I think the best compromise then is to remove this prompt and make english the (only) default selected language, so we can have the same behavior, 1 less prompt and no misleading as i18n cannot be really disabled :)

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (sinedied) [reply README.md]: I don't think i18n can be qualified as a feature module, as it's a transverse utility without any business value and thus belongs to core/shared module in regard to Angular style guide.

Most projects will adapt the base template anyways, and I would prefer keeping it as close as possible to the Angular style guide.

File: generators/app/templates/tsconfig.json (hunk 10 above)
Comment (sinedied) [reply generators/app/templates/tsconfig.json]: Yes I understand the idea behind it, but it should not be forced by default as it's not supported properly by some IDEs and online editors. I'm for keeping this under an optional flag, but that should be done in a separate PR.

File: generators/app/prompts.js (hunk 13 above)
Comment (snowyu) [reply generators/app/prompts.js]: Clear clever.

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (snowyu) [reply README.md]: First the translation is the i18n's business.

The i18n service and components together could be easy added or remove to/from an app.
Or you should search the translation feature on the core module and others on the share module.
And even It could be extract as separate library(the only problem is howto be Independent the UI).  But It can be reused on different ionic/md/bootstrap application(Not just as a template application, but also as a library).

The `core` and `share` module should put the services and components only avaiable on the app.
They can not be ported to another app.


eg, I have a Main App and many Sub-App work/compile together(they used/shared the same UI).
The each app could have to a piece of i18n code If the i18n is not a library.

[the remark this answers is not in the document]
Comment (sinedied) [reply README.md]: I understand the rationale behind it, but I still think it would be better to keep it under existing core/shared module for the *default* generated template.
We try to stand behind the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) whenever we can, and IMHO adding a i18n module here goes too far for a base sample code aiming to be used in any app.

I agree your separation may be useful in some cases as you stated, but the opposite can also be said: in my company apps, the i18n services/components are heavily customized for each app and therefore it does not make sense to separate them from other transverse services.

And if you plan to share some services/components between apps, you should better make it a separate library, you can even use [Nx extensions](https://nrwl.io/nx) to make it in a same repo for your app/sub-apps.

If you want to make this separation the default for you own apps you can still do this by creating an add-on generator, but I don't think it's a suitable default for everyone especially as the Angular Style Guide specifies that such transverse services/components should be put in core/shared.

[the remark this answers is not in the document]
Comment (snowyu) [reply README.md]: I think this is a demo to teach guys : how to reuse and organize code in an application?

* Application Layer
  * Core/Shared
* Internal Library Layer
* External Library Layer 

BTW The NX is too heavey for me to KISS.

**Comment by bursauxa** [comment]
I would very much like this PR to move forward, as I believe the language selector adds good value to ngx-rocket. There are still two blocking issues :

- merge with #369 ;
- use of `@app` and similar shortcuts.

Regarding the first issue, I merged the Ionic changes to the best of my ability without a dev rig (6dc0cba). Merge through the GitHub UI is very YOLO for this kind of thing 😨. Hence testing will be required, especially since I do not know much about what Ionic should look like so maybe someone more knowledgeable should have a look.

Regarding the second issue, I'm progressively feeling inclined towards integrating the use of the `@app`-like shortcuts. Admittedly this would have been better as a separate PR ; on the other hand, since this PR uses the mechanism itself (with `@i18n`), it would have been blocked since we have no PR addressing the shortcuts alone. I can also understand why @snowyu would not want to deconstruct his previous work ; this is unpaid labor, after all.

So my point is the following : I believe that we should include the shortcuts regardless of other considerations ; having them in this PR is not ideal but better than having neither them neither the language selector ; having an option to change `@` with other characters (as mentioned by @sinedied in #277) is useful but not mandatory as it can be a future evolution.

Sorry for the wall of text 🏰 ! Now I'd like to ask for your inputs and see how we can move forward.

**Comment by sinedied** [comment]
Whoops, this PR was closed automatically because I removed the base branch, my bad 😭 
it should be rebased on master now that v6 is merged.

I 👍 @bursauxa comments (sorry for the long delay), we can always update the shortcuts thing later on.

**Comment by sinedied** [comment]
It's been a while, but I finally updated and adapted the code for the latest version (see #537).
Kudos to @snowyu for this, it should have been merged a while ago!
